src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,202 @@
+            exist.
+        @rtype: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pidString = b""
+        try:
+            for pidString in self.filePath.open():
+                break
+        except OSError as e:
+            if e.errno == errno.ENOENT:  # No such file
+                return None
+            raise
+
+        return int(pidString)
+
+
+    def write(self, pid=None):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.  If C{None}, store the currently running
+            process ID.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+        if pid is None:
+            pid = getpid()
+        else:
+            pid = int(pid)
+
+        self.filePath.setContent(self.format(pid=pid))
+
+
+    def remove(self):
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+        self.filePath.remove()
+
+
+    def isRunning(self):
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pid = self.read()
+        if pid is None:
+            return False
+
+        try:
+            kill(pid, 0)
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process
+                self._log.info("Removing stale PID file: {log_source}")
+                self.remove()
+                return False
+            elif e.errno == errno.EPERM:  # Not permitted to kill
+                return True
+            else:
+                raise
+        else:
+            return True
+
+
+    def __enter__(self):
+        """
+        Enter a context using this PIDFile.
+
+        Writes the PID file with the PID of the running process.
+
+        @raise AlreadyRunningError: A process corresponding to the PID in this
+            PID file is already running.
+        """
+        if self.isRunning():
+            raise AlreadyRunningError()
+        self.write()
+        return self
+
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Exit a context using this PIDFile.
+
+        Removes the PID file.
+        """
+        self.remove()
+
+
+
+class NonePIDFile(PIDFile):
+    """
+    PID file implementation that does nothing.
+
+    This is meant to be used as a "active None" object in place of a PID file
+    when no PID file is desired.
+    """
+
+    def __init__(self):
```

Comment:
Doesn't this no-argument initializer mean that `NonePIDFile` doesn't implement the `IPIDFile` interface?  If so, is this OK because we prefer to verify that an object provides an interface to verifying that its class implements it?

If this should take an argument to ensure that the class implements `NonePIDFile`, is a default value of `None` acceptable and a note in the doc string that the argument is ignored?